File: 01-search-algorithms/examples.py
"""
BFS и DFS на графе, заданном списком смежности.
Запуск: python examples.py
"""

from collections import deque


def bfs(graph: dict[int, list[int]], start: int) -> list[int]:
    """Возвращает порядок обхода в ширину от start."""
    visited = {start}
    queue = deque([start])
    order = []
    while queue:
        v = queue.popleft()
        order.append(v)
        for u in graph.get(v, []):
            if u not in visited:
                visited.add(u)
                queue.append(u)
    return order


def bfs_shortest_path(graph: dict[int, list[int]], start: int, goal: int) -> list[int] | None:
    """Кратчайший путь в невзвешенном графе. None — если пути нет."""
    if start == goal:
        return [start]
    parents = {start: None}
    queue = deque([start])
    while queue:
        v = queue.popleft()
        for u in graph.get(v, []):
            if u in parents:
                continue
            parents[u] = v
            if u == goal:
                path = [u]
                while parents[path[-1]] is not None:
                    path.append(parents[path[-1]])
                return list(reversed(path))
            queue.append(u)
    return None


def dfs_recursive(graph: dict[int, list[int]], start: int) -> list[int]:
    """Рекурсивный DFS. Осторожно: глубокая рекурсия → RecursionError."""
    visited: set[int] = set()
    order: list[int] = []

    def go(v: int) -> None:
        visited.add(v)
        order.append(v)
        for u in graph.get(v, []):
            if u not in visited:
                go(u)

    go(start)
    return order


def dfs_iterative(graph: dict[int, list[int]], start: int) -> list[int]:
    """Итеративный DFS — без риска переполнить стек."""
    visited = {start}
    stack = [start]
    order = []
    while stack:
        v = stack.pop()
        order.append(v)
        # reversed чтобы порядок совпадал с рекурсивной версией
        for u in reversed(graph.get(v, [])):
            if u not in visited:
                visited.add(u)
                stack.append(u)
    return order


def has_cycle_undirected(graph: dict[int, list[int]]) -> bool:
    """Поиск цикла в неориентированном графе через DFS."""
    visited: set[int] = set()

    def go(v: int, parent: int) -> bool:
        visited.add(v)
        for u in graph.get(v, []):
            if u not in visited:
                if go(u, v):
                    return True
            elif u != parent:
                return True
        return False

    for v in graph:
        if v not in visited and go(v, -1):
            return True
    return False


if __name__ == "__main__":
    #     1 - 2
    #     |   |
    #     3 - 4 - 5
    g = {
        1: [2, 3],
        2: [1, 4],
        3: [1, 4],
        4: [2, 3, 5],
        5: [4],
    }
    print("BFS от 1:", bfs(g, 1))
    print("DFS (рекурсия) от 1:", dfs_recursive(g, 1))
    print("DFS (итеративный) от 1:", dfs_iterative(g, 1))
    print("Кратчайший путь 1 -> 5:", bfs_shortest_path(g, 1, 5))
    print("Есть ли цикл:", has_cycle_undirected(g))

File: 01-search-algorithms/examples.go
// BFS и DFS на графе, заданном списком смежности.
// Запуск: go run examples.go
package main

import "fmt"

type Graph map[int][]int

func BFS(g Graph, start int) []int {
	visited := map[int]bool{start: true}
	queue := []int{start}
	order := []int{}
	for len(queue) > 0 {
		v := queue[0]
		queue = queue[1:]
		order = append(order, v)
		for _, u := range g[v] {
			if !visited[u] {
				visited[u] = true
				queue = append(queue, u)
			}
		}
	}
	return order
}

// BFSShortestPath возвращает кратчайший путь в невзвешенном графе.
// nil — если пути нет.
func BFSShortestPath(g Graph, start, goal int) []int {
	if start == goal {
		return []int{start}
	}
	parents := map[int]int{start: -1}
	queue := []int{start}
	for len(queue) > 0 {
		v := queue[0]
		queue = queue[1:]
		for _, u := range g[v] {
			if _, seen := parents[u]; seen {
				continue
			}
			parents[u] = v
			if u == goal {
				path := []int{u}
				for parents[path[len(path)-1]] != -1 {
					path = append(path, parents[path[len(path)-1]])
				}
				// reverse
				for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
					path[i], path[j] = path[j], path[i]
				}
				return path
			}
			queue = append(queue, u)
		}
	}
	return nil
}

func DFSRecursive(g Graph, start int) []int {
	visited := map[int]bool{}
	order := []int{}
	var go_ func(v int)
	go_ = func(v int) {
		visited[v] = true
		order = append(order, v)
		for _, u := range g[v] {
			if !visited[u] {
				go_(u)
			}
		}
	}
	go_(start)
	return order
}

func DFSIterative(g Graph, start int) []int {
	visited := map[int]bool{start: true}
	stack := []int{start}
	order := []int{}
	for len(stack) > 0 {
		v := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		order = append(order, v)
		neigh := g[v]
		for i := len(neigh) - 1; i >= 0; i-- {
			u := neigh[i]
			if !visited[u] {
				visited[u] = true
				stack = append(stack, u)
			}
		}
	}
	return order
}

func main() {
	g := Graph{
		1: {2, 3},
		2: {1, 4},
		3: {1, 4},
		4: {2, 3, 5},
		5: {4},
	}
	fmt.Println("BFS от 1:", BFS(g, 1))
	fmt.Println("DFS (рекурсия) от 1:", DFSRecursive(g, 1))
	fmt.Println("DFS (итеративный) от 1:", DFSIterative(g, 1))
	fmt.Println("Кратчайший путь 1 -> 5:", BFSShortestPath(g, 1, 5))
}
